File: Exercise4.py
# Time Complexity : O(nlogn)
# Space Complexity : O(n)
# Did this code successfully run on Leetcode : N/A
# Any problem you faced while coding this : No


# Python program for implementation of MergeSort 

def merge(arr, low, mid, high):
  temp = []
  left = low
  right = mid + 1
  while left <= mid and right <= high:
    if arr[left] <= arr[right]:
      temp.append(arr[left])
      left = left + 1
    else:
      temp.append(arr[right])
      right = right + 1
  while left <= mid:
    temp.append(arr[left])
    left = left + 1
  while right <= high:
    temp.append(arr[right])
    right = right + 1
  for i in range(low, high+1):
    arr[i] = temp[i-low]



def ms(arr, low, high):
  if low < high:
    mid = (low+high)//2
    ms(arr, low, mid)
    ms(arr, mid+1, high)
    merge(arr, low, mid, high)

def mergeSort(arr):

  #write your code here
  ms(arr, 0, len(arr)-1)

# Code to print the list 
def printList(arr): 

    #write your code here
    print(arr)

# driver code to test the above code 
if __name__ == '__main__': 
    arr = [12, 11, 13, 5, 6, 7]  
    print ("Given array is", end="\n")  
    printList(arr) 
    mergeSort(arr) 
    print("Sorted array is: ", end="\n") 
    printList(arr)

File: Exercise_4.swift
import Cocoa

// Time Complexity -  O(nlogn)
// Space Complexity - O(n)
func mergeSortArray(_ nums: [Int]) -> [Int] {
        var array = nums
        
        mergeSort(&array, 0, array.count-1)
        
        return array
    }
    
    func mergeSort(_ array: inout [Int], _ left: Int, _ right: Int) {
        if (left >= right) {return}
        
        let mid: Int = left + (right-left)/2
        
        mergeSort(&array, left, mid)
        mergeSort(&array, mid+1, right)
        
        merge(&array, left, right)
    }
    
    func merge(_ array: inout[Int], _ left: Int, _ right: Int) {
        let leftEnd: Int = left+(right-left)/2
        var l = left
        var r = leftEnd+1
        let rightEnd = right
        
        var helper:[Int] = Array()
        
        while (l <= leftEnd && r <= rightEnd) {
            if array[l] <= array[r] {
                helper.append(array[l])
                l += 1
            } else {
                helper.append(array[r])
                r += 1
            }
        }
        
        while (l <= leftEnd) {
            helper.append(array[l])
            l += 1
        }
        
        while (r <= rightEnd) {
            helper.append(array[r])
            r += 1
        }
        
        for i in 0...(right-left) {
            array[left+i] = helper[i]
        }
    }


let array = [10, 20, 1, 4, 100, 2, 15, 56]
print(mergeSortArray(array))


